### PR TITLE
Debug ceph storage by defining K8sResources to debug and logging each

### PR DESCRIPTION
**Pull request by addyess**
This is a refactor of `_debug_ceph_storage` so i can reuse a basic structure to debug dns tests

### CODE REVIEW

File: ci.bash [hunk 1]
```diff
@@ -180,7 +180,7 @@ function test::capture
             -j snap.cdk-addons* \
             -m "$JUJU_CONTROLLER:$JUJU_MODEL"
     fi
-    tar -cvzf artifacts.tar.gz ci.log _out meta juju-crashdump* report.* failures* || true
+    tar -cvzf artifacts.tar.gz ci.log _out meta juju-crashdump* report.* failures* logs/ || true
```
Comment (addyess) [inline ci.bash:183]: including the `logs/` directory in the zipped artifact

File: jobs/integration/conftest.py [hunk 2]
```diff
@@ -470,14 +471,30 @@ def skip_by_model(request, model):
...
+        if "mode" not in kwargs:
+            # defaults to writing log files if mode not specified
+            kwargs["mode"] = "w"
+        with path.open(**kwargs) as fp:
+            log(f"Logging to {path}... ")
+            yield fp
+
+    yield _open
+
+
```
Comment (addyess) [inline jobs/integration/conftest.py:497]: this fixture yields a method which can be used like python's `open` in a with context to write log files in a directory based on the name of the current running test.

File: jobs/integration/utils.py [hunk 3]
```diff
@@ -464,12 +468,13 @@ async def validate_storage_class(model, sc_name, test_name):
...
-            pods=[f"{sc_name}-write-test", f"{sc_name}-read-test"],
-        )
-
+    except:
+        # bare except intentional to debug any failure
+        if debug_open:
+            await _debug_storage_class(
+                debug_open, test_name, sc_name, provisioner, model
+            )
+        raise
```
Comment (addyess) [inline jobs/integration/utils.py:477]: if we catch ANY error -- let's debug that error but don't continue the testing -- whichever exception needs to be re-raised so the test doesn't accidentally PASS

File: jobs/integration/utils.py [hunk 4]
```diff
@@ -496,37 +501,40 @@ async def validate_storage_class(model, sc_name, test_name):
         )
 
 
-async def _debug_ceph_storage(control_plane, pods):
-    log.error("Gathering Ceph logs from Cluster")
+async def _debug_storage_class(debug_open, test_name, sc_name, provisioner, model):
+    class _Call:
+        def __init__(self, *args, **kwds):
+            self.args, self.kwds = args, kwds
```
Comment (addyess) [inline jobs/integration/utils.py:507]: class to hold variable args and kwargs

File: jobs/integration/utils.py [hunk 5]
```diff
@@ -496,37 +501,40 @@ async def validate_storage_class(model, sc_name, test_name):
...
+    provisioner_pods = []
+    if provisioner:
+        result = await kubectl(
+            model,
+            "get",
+            "pods",
+            A=True,
+            l=f"app={provisioner}",
+            o=r"""jsonpath='{range .items[*]}{@..metadata.namespace}{" "}{@..metadata.name}{"\n"}{end}'""",
+        )
```
Comment (addyess) [inline jobs/integration/utils.py:521]: this `jsonpath` formats the pods as `<pod-namespace> <pod-name` per line of output

using that, we can parse the output lines and look up the logs of the provisioner further down

**Review by mateoflorido — APPROVED** [review]
LGTM ✨

**Pull request merged**